ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/P2TargetUtils.java

```diff
@@ -1418,13 +1422,36 @@ private Collection<URI> getArtifactRepositories(ITargetDefinition target) throws
 			// get all the artifact repos we know in the manager currently
 			result.addAll(Arrays.asList(manager.getKnownRepositories(IRepositoryManager.REPOSITORIES_ALL)));
 
+			try {
+				result.addAll(getSharedBundlePools());
+			} catch (IOException e) {
+				//$FALL-THROUGH$
+			}
+
 			// Add in the IDE profile bundle pool and all known workspaces
 			findProfileRepos(result);
 			findWorkspaceRepos(result);
 		}
 		return result;
 	}
 
+	@SuppressWarnings("nls")
+	private Set<URI> getSharedBundlePools() throws IOException {
+		Set<URI> result = new LinkedHashSet<>();
+		if ("true".equals(System.getProperty("pde.usePoolsInfo", "true"))) {
+			Path bundlePools = Path.of(System.getProperty("user.home"), ".p2/pools.info");
+			if (Files.isRegularFile(bundlePools)) {
+				for (String poolLocation : Files.readAllLines(bundlePools, Charset.defaultCharset())) {
+					Path poolLocationPath = Path.of(poolLocation);
+					if (Files.isDirectory(poolLocationPath)) {
+						result.add(poolLocationPath.toUri());
+					}
+				}
+			}
+		}
+		return result;
+	}
```

Comment:
Using streams this could be a little bit more compact.

Furthermore I wonder if it would be good to always use UTF-8 to write that file?

```suggestion
	private Collection<URI> getSharedBundlePools() throws IOException {
		if ("true".equals(System.getProperty("pde.usePoolsInfo", "true"))) {
			Path bundlePools = Path.of(System.getProperty("user.home"), ".p2/pools.info");
			if (Files.isRegularFile(bundlePools)) {
				return Files.lines(bundlePools, Charset.defaultCharset()) //
						.map(Path::of).filter(Files::isDirectory) //
						.map(Path::toUri).toList();
			}
		}
		return List.of();
	}
```

Reply:
I forgot the try-with-resources. Updated the suggestion.

Reply:
I wish we did always use UTF-8 to write that file. 😱   

But I looked at the implementation and we use new FileReader(file).  😭 

It's rather tragic...

Reply:
> It's rather tragic...

Indeed. Especially that with [JEP 400: UTF-8 by Default](https://openjdk.org/jeps/400) delivered in Java 18 UTF-8 will become the default default.
So maybe it would be good to have a migration path anyways (maybe write the encoding as comment in the first line and re-read again if it has not written accordingly).

Reply:
Indeed. I’m concerned how older versions can handle newly written changes. What a nightmare on Windows.

Reply:
> Indeed. I’m concerned how older versions can handle newly written changes. What a nightmare on Windows.

Maybe this helps:
- https://softwareengineering.stackexchange.com/a/187174
- https://learn.microsoft.com/en-us/windows/win32/intl/using-byte-order-marks?redirectedfrom=MSDN

Or the poor man's variant of having a initial comment indicating the encoding. The absence could be treated as Windows default (IIRC `ISO_8859_1`).

Reply:
If I change the serialization form in any way not recognized by older versions of the libraries, I think I'm going to get bitten.   Here is a thought...

I write a file that's effectively created and initialized one time.  It contains the encoding used for all the other files.  It's written in UTF-8.  Older runtimes will just continue to use the default encoding as before, not looking at this file.  Newer runtimes will create the file if it doesn't exist and subsequently will use the encoding in that file from then on.  When eventually Java 21 changes the default to UTF-8 the newer runtimes will continue to use the recorded encoding that was used for the existing files.  In the future, brand-new shared agent would just record and use UTF-8 everywhere.

I've prototyped that in Oomph and the code here for PDE could be made safely compatible with that approach as follows:

![image](https://github.com/eclipse-pde/eclipse.pde/assets/208716/a774ec75-cd5d-4a25-ade5-5cd735c9b122)

What do you think?

Reply:
I did some extensive testing with a user home folder `D:\Users\tëst33\.p2\pool` with special characters.   The encoding used by default right now is `windows-1252`.  I tested that if `pools.info` is read with UTF-8 that it fails to determine the pool locations correctly.   With the extended logic, even if the system default encoding is UTF-8, it reads the encoding from `encoding.info` and then reads the `pools.info` correctly...

Reply:
I have a commit ready for Oomph to create the `encoding.info` and to use it as well to future proof against changes to the default encoding.  Of course the code here for PDE works with or without the `encoding.info`, as long as PDE is running with the same default encoding as is currently the case with Java 17 (and of course if it isn't it will just quietly not use the shared bundle pool).

Reply:
@merks I think you can quite easily hand this, simply read the file as byte array, then decode it at "windows-1252" and "utf-8", the one where you find a valid file is the one with correct encoding (in some case simply both will be the same)...

Reply:
The default encoding on Windows in Japan is Shift-JIS I think.  In China I think it's MS936.  I really don't know all the defaults in various countries.  I'd really like to avoid assumptions.  Furthermore, there are a number of files being read **and written** by this infrastructure in Oomph so I'd like to ensure they all continue to work as they do today.   So ideal to my thinking is to know the encoding and use exactly that correct encoding when reading and when writing so that no existing installations will suddenly just stop work...

Reply:
Good point, whatever solution is used, I think it would be good to add that reading support to P2 as suggested here:
- https://github.com/eclipse-equinox/p2/issues/373

and just consume this then from Oomph, PDE (and after the release from Tycho) as well, also it would be good to write the encoding file if it is not present yet so it is ensured the current encoding is recorded for status-quo.

Reply:
Given it just returns a list of URIs, it could be added almost anywhere.  Any specific suggestion in that regard?

Reply:
If we want to keep it internal I think `org.eclipse.equinox.internal.p2.repository.helpers.RepositoryHelper` would be suitable, otherwise I would just create a new class `P2Pool` in package `org.eclipse.equinox.p2.repository.artifact`.

Reply:
The latest changes have moved the logic to p2 and reuse it in PDE.